src/main/java/org/inferred/freebuilder/processor/property/BiMapProperty.java

```diff
@@ -0,0 +1,406 @@
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.inferred.freebuilder.processor.property;
+
+import static org.inferred.freebuilder.processor.BuilderMethods.clearMethod;
+import static org.inferred.freebuilder.processor.BuilderMethods.getter;
+import static org.inferred.freebuilder.processor.BuilderMethods.mutator;
+import static org.inferred.freebuilder.processor.BuilderMethods.putAllMethod;
+import static org.inferred.freebuilder.processor.BuilderMethods.putMethod;
+import static org.inferred.freebuilder.processor.model.ModelUtils.erasesToAnyOf;
+import static org.inferred.freebuilder.processor.model.ModelUtils.maybeDeclared;
+import static org.inferred.freebuilder.processor.model.ModelUtils.maybeUnbox;
+import static org.inferred.freebuilder.processor.model.ModelUtils.overrides;
+import static org.inferred.freebuilder.processor.model.ModelUtils.upperBound;
+import static org.inferred.freebuilder.processor.property.MergeAction.appendingToCollections;
+import static org.inferred.freebuilder.processor.source.FunctionalType.consumer;
+import static org.inferred.freebuilder.processor.source.FunctionalType.functionalTypeAcceptedByMethod;
+
+import com.google.common.collect.*;
+
+import org.inferred.freebuilder.processor.Datatype;
+import org.inferred.freebuilder.processor.Declarations;
+import org.inferred.freebuilder.processor.excerpt.CheckedBiMap;
+import org.inferred.freebuilder.processor.source.Excerpt;
+import org.inferred.freebuilder.processor.source.FunctionalType;
+import org.inferred.freebuilder.processor.source.SourceBuilder;
+import org.inferred.freebuilder.processor.source.Type;
+import org.inferred.freebuilder.processor.source.Variable;
+
+import java.util.*;
+
+import javax.lang.model.element.TypeElement;
+import javax.lang.model.type.DeclaredType;
+import javax.lang.model.type.TypeMirror;
+import javax.lang.model.util.Elements;
+import javax.lang.model.util.Types;
+
+/**
+ * {@link PropertyCodeGenerator} providing fluent methods for {@link BiMap} properties.
+ */
+class BiMapProperty extends PropertyCodeGenerator {
+
+  static class Factory implements PropertyCodeGenerator.Factory {
+
+    @Override
+    public Optional<BiMapProperty> create(Config config) {
+      Property property = config.getProperty();
+      DeclaredType type = maybeDeclared(property.getType()).orElse(null);
+      if (!erasesToAnyOf(
+          type,
+          EnumHashBiMap.class,
+          EnumBiMap.class,
+          HashBiMap.class,
```

Comment:
Only things which can be immutable (BiMap and ImmutableBiMap) should be here, just as HashMap isn't in the list for MapProperty.